source/time_loop_mgrit.h

```diff
@@ -0,0 +1,210 @@
+        /**
+         * @copydoc HyperbolicSystem
+         */
+        using HyperbolicSystem = typename Description::HyperbolicSystem;
+
+        /**
+         * @copydoc ParabolicSystem
+         */
+        using ParabolicSystem = typename Description::ParabolicSystem;
+
+        /**
+         * @copydoc HyperbolicSystem::View
+         */
+        using HyperbolicSystemView =
+            typename Description::HyperbolicSystem::template View<dim, Number>;
+
+        /**
+         * @copydoc HyperbolicSystem::problem_dimension
+         */
+        static constexpr unsigned int problem_dimension =
+            HyperbolicSystemView::problem_dimension;
+
+        /**
+         * @copydoc HyperbolicSystem::n_precomputed_values
+         */
+        static constexpr unsigned int n_precomputed_values =
+            HyperbolicSystemView::n_precomputed_values;
+
+
+        /**
+         * @copydoc OfflineData::scalar_type
+         */
+        using scalar_type = typename OfflineData<dim, Number>::scalar_type;
+
+        /**
+         * Typedef for a MultiComponentVector storing the state U.
+         */
+        using vector_type = MultiComponentVector<Number, problem_dimension>;
+
+        /**
+         * Typedef for a MultiComponentVector storing precomputed values.
+         */
+        using precomputed_type = MultiComponentVector<Number, n_precomputed_values>;
+
+        /**
+         * Constructor.
+         */
+        TimeLoopMgrit(const MPI_Comm &mpi_comm,
+                      const LevelStructures<Description, dim, Number> &ls,
+                      const Number initial_time,
+                      const Number final_time);
+
+        /**
+         * Run the high-level time loop.
+         */
+        void run();
+
+        void run_with_initial_data(const vector_type &U);
+
+        //returns the solution state U, if the size is greater than one.
+        vector_type get_U();
+
+      protected:
+        /**
+         * @name Private methods for run()
+         */
+        //@{
+
+        void compute_error(const vector_type &U, Number t);
+
+        void output(const vector_type &U,
+            const std::string &name,
+            Number t,
+            unsigned int cycle);
+
+        void print_parameters(std::ostream &stream);
+        void print_mpi_partition(std::ostream &stream);
+        void print_memory_statistics(std::ostream &stream);
+        void print_timers(std::ostream &stream);
+        void print_throughput(unsigned int cycle,
+            Number t,
+            std::ostream &stream,
+            bool final_time = false);
+
+        void print_info(const std::string &header);
+        void print_head(const std::string &header,
+            const std::string &secondary,
+            std::ostream &stream);
+
+        void print_cycle_statistics(unsigned int cycle,
+            Number t,
+            unsigned int output_cycle,
+            bool write_to_logfile = false,
+            bool final_time = false);
```

Comment:
Is there a purpose in separating all these `print` functions? If you all call them at the same time, then I would just merge them into one.